src/libraries/System.Net.Security/tests/UnitTests/NTAuthenticationTests.cs

```diff
@@ -0,0 +1,119 @@
+            //   NTLMSSP_NEGOTIATE_NTLM
+            //   NTLMSSP_NEGOTIATE_SEAL
+            //   NTLMSSP_NEGOTIATE_SIGN
+            //   NTLMSSP_NEGOTIATE_OEM
+            //   NTLMSSP_NEGOTIATE_UNICODE
+            // Domain: (empty) (should be "Domain" but the fake server doesn't check)
+            // Workstation: (empty) (should be "COMPUTER" but the fake server doesn't check)
+            // Version: 6.1.7600 / 15
+            byte[] negotiateBlob = Convert.FromHexString("4e544c4d535350000100000033828ae2000000000000000000000000000000000601b01d0000000f");
+            byte[]? challengeBlob = fakeNtlmServer.GetOutgoingBlob(negotiateBlob);
+
+            // CHALLENGE_MESSAGE from 4.2.4.3 Messages
+            byte[] expectedChallengeBlob = Convert.FromHexString(
+                "4e544c4d53535000020000000c000c003800000033828ae20123456789abcdef" +
+                "00000000000000002400240044000000060070170000000f5300650072007600" +
+                "6500720002000c0044006f006d00610069006e0001000c005300650072007600" +
+                "6500720000000000");
+            Assert.Equal(expectedChallengeBlob, challengeBlob);
+
+            // AUTHENTICATE_MESSAGE from 4.2.4.3 Messages
+            byte[] authenticateBlob = Convert.FromHexString(
+                "4e544c4d5353500003000000180018006c00000054005400840000000c000c00" +
+                "480000000800080054000000100010005c00000010001000d8000000358288e2" +
+                "0501280a0000000f44006f006d00610069006e00550073006500720043004f00" +
+                "4d005000550054004500520086c35097ac9cec102554764a57cccc19aaaaaaaa" +
+                "aaaaaaaa68cd0ab851e51c96aabc927bebef6a1c010100000000000000000000" +
+                "00000000aaaaaaaaaaaaaaaa0000000002000c0044006f006d00610069006e00" +
+                "01000c005300650072007600650072000000000000000000c5dad2544fc97990" +
+                "94ce1ce90bc9d03e");
+            byte[]? empty = fakeNtlmServer.GetOutgoingBlob(authenticateBlob);
+            Assert.Null(empty);
+            Assert.True(fakeNtlmServer.IsAuthenticated);
+            Assert.False(fakeNtlmServer.IsMICPresent);
+        }
+
+        [ConditionalFact(nameof(IsNtlmInstalled))]
+        public void NtlmCorrectExchangeTest()
+        {
+            FakeNtlmServer fakeNtlmServer = new FakeNtlmServer(s_testCredentialRight);
+            NTAuthentication ntAuth = new NTAuthentication(
+                isServer: false, "NTLM", s_testCredentialRight, "HTTP/foo",
+                ContextFlagsPal.Connection | ContextFlagsPal.InitIntegrity, null);
+
+            DoNtlmExchange(fakeNtlmServer, ntAuth);
+
+            Assert.True(fakeNtlmServer.IsAuthenticated);
+            // NTLMSSP on Linux doesn't send the MIC and sends incorrect SPN (drops the service prefix)
+            if (!OperatingSystem.IsLinux())
+            {
+                Assert.True(fakeNtlmServer.IsMICPresent);
+                Assert.Equal("HTTP/foo", fakeNtlmServer.ClientSpecifiedSpn);
```

Comment:
[gss-ntlmssp seems to throw away the service name](https://github.com/gssapi/gss-ntlmssp/blob/9b6493b1419d615e347d67c636fd238aa4d0f780/src/gss_names.c#L297-L319). I think that's a bug in their implementation. Neither Windows nor macOS seems to do that.

Reply:
Filed a bug: https://github.com/gssapi/gss-ntlmssp/issues/63